opentelemetry-api/src/opentelemetry/trace/propagation/textmap.py

```diff
@@ -18,9 +18,71 @@
 from opentelemetry.context.context import Context
 
 TextMapPropagatorT = typing.TypeVar("TextMapPropagatorT")
+CarrierValT = typing.Union[typing.List[str], str]
 
 Setter = typing.Callable[[TextMapPropagatorT, str, str], None]
-Getter = typing.Callable[[TextMapPropagatorT, str], typing.List[str]]
+
+
+class Getter(typing.Generic[TextMapPropagatorT]):
+    """This class implements a Getter that enables extracting propagated
+    fields from a carrier
+
+    """
+
+    def get(self, carrier: TextMapPropagatorT, key: str) -> typing.List[str]:
+        """Function that can retrieve zero
+        or more values from the carrier. In the case that
+        the value does not exist, returns an empty list.
+
+        Args: carrier: and object which contains values that are used to
+        construct a Context. This object must be paired with an appropriate
+        getter which understands how to extract a value from it.
+        key: key of a field in carrier. Returns: first value of the
+        propagation key or an empty list if the key doesn't exist.
+        """
+        raise NotImplementedError()
+
+    def keys(self, carrier: TextMapPropagatorT) -> typing.List[str]:
+        """Function that can retrieve all the keys in a carrier object.
+
+        Args:
+            carrier: and object which contains values that are
+                used to construct a Context. This object
+                must be paired with an appropriate getter
+                which understands how to extract a value from it.
+        Returns:
+            list of keys from the carrier.
+        """
+        raise NotImplementedError()
+
+
+class DictGetter(Getter[typing.Dict[str, CarrierValT]]):
+    def get(
+        self, carrier: typing.Dict[str, CarrierValT], key: str
+    ) -> typing.List[str]:
+        val = carrier.get(key, None)
+        if not val:
+            return []
+        return val if isinstance(val, typing.List) else [val]
```

Comment:
would it  be better to have this:

```python
    if not isinstance(val, str) and isinstance(value, Iterable):
        return list(val)
    return [val]
```

This way tuples or any other collection values would work as well.